### PR TITLE
K8SPSMDB-455: Fix downgrade

### PR DESCRIPTION
**Pull request by blez**
[K8SPSMDB-455](https://jira.percona.com/browse/K8SPSMDB-455) [Powered by Pull Request Badge](https://pullrequestbadge.com/?utm_medium=github&utm_source=percona&utm_campaign=badge_info)

### CODE REVIEW

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |

commit: https://github.com/percona/percona-server-mongodb-operator/pull/654/commits/d87d603d42606d3aa5702099ec34ee75ba13e268
image: `perconalab/percona-server-mongodb-operator:PR-654-d87d603d`

**Review by qJkee — DISMISSED** [review]

File: pkg/controller/perconaservermongodb/version.go [hunk 1]
```diff
@@ -189,9 +189,27 @@ func majorUpgradeRequested(cr *api.PerconaServerMongoDB, fcv string) (UpgradeReq
 		return UpgradeRequest{true, apply, ver}, nil
 	}
 
-	can, err := canUpgradeVersion(fcv, ver)
+	mongoVer, err := v.NewSemver(cr.Status.MongoVersion)
 	if err != nil {
-		return UpgradeRequest{false, "", ""}, errors.Wrap(err, "can't upgrade")
+		return UpgradeRequest{false, "", ""}, errors.Wrap(err, "faied to make semver")
```
Comment (delgod) [inline pkg/controller/perconaservermongodb/version.go:194]:
```suggestion
		return UpgradeRequest{false, "", ""}, errors.Wrap(err, "failed to make semver")
```

File: pkg/controller/perconaservermongodb/version.go [hunk 2]
```diff
@@ -189,9 +189,27 @@ func majorUpgradeRequested(cr *api.PerconaServerMongoDB, fcv string) (UpgradeReq
...
+	}
+
+	newMM := MajorMinor(newVer)
+	mongoMM := MajorMinor(mongoVer)
+
+	can := false
+
+	if newMM > mongoMM {
```
Comment (delgod) [inline pkg/controller/perconaservermongodb/version.go:202]: now we have comparison logic in two places :( one in `canUpgradeVersion` and another one in majorUpgradeRequested 😞 
can we handle logic in one place?

**Review by delgod — CHANGES_REQUESTED** [review]

File: pkg/controller/perconaservermongodb/version.go (hunk 2 above)
Comment (blez) [reply pkg/controller/perconaservermongodb/version.go:202]: But the comparisons are different.
Here we check if user want to update or downgrade.
There we check if we can upgrade from one version to another.

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| init-deploy | passed |
| storage | passed |
| one-pod | passed |
| upgrade-consistency | passed |
| monitoring-2-0 | passed |
| self-healing | passed |
| demand-backup | passed |
| limits | passed |
| arbiter | passed |
| scaling | passed |
| scheduled-backup | passed |
| operator-self-healing | passed |
| security-context | passed |
| service-per-pod | passed |
| demand-backup-sharded | failed |
| rs-shard-migration | passed |
| liveness | passed |
| smart-update | passed |
| version-service | passed |
| users | passed |
| upgrade | passed |
| data-sharded | passed |
| upgrade-sharded | passed |
| pitr | passed |

commit: https://github.com/percona/percona-server-mongodb-operator/pull/654/commits/d87d603d42606d3aa5702099ec34ee75ba13e268
image: `perconalab/percona-server-mongodb-operator:PR-654-d87d603d`

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| init-deploy | passed |
| storage | passed |
| one-pod | passed |
| upgrade-consistency | passed |
| monitoring-2-0 | passed |
| self-healing | passed |
| demand-backup | passed |
| limits | passed |
| scaling | passed |
| arbiter | passed |
| scheduled-backup | passed |
| operator-self-healing | passed |
| security-context | passed |
| service-per-pod | passed |
| demand-backup-sharded | failed |
| rs-shard-migration | passed |
| liveness | passed |
| smart-update | passed |
| version-service | passed |
| users | passed |
| upgrade | passed |
| data-sharded | passed |
| upgrade-sharded | passed |
| pitr | passed |

commit: https://github.com/percona/percona-server-mongodb-operator/pull/654/commits/be6c565474ea4d0e998eebf66070432996b0889e
image: `perconalab/percona-server-mongodb-operator:PR-654-be6c5654`

**Review by dAdAbird — COMMENTED** [review]
See my comment.
Plus, know it is out of scope for this PR, but shouldn't "ToLower()" and other casting be a part of `v1.CheckAndSetDefaults`?

File: pkg/controller/perconaservermongodb/version.go [hunk 3]
```diff
@@ -189,9 +189,27 @@ func majorUpgradeRequested(cr *api.PerconaServerMongoDB, fcv string) (UpgradeReq
...
+		can, err = canUpgradeVersion(fcv, ver)
+		if err != nil {
+			return UpgradeRequest{false, "", ""}, errors.Wrap(err, "can't upgrade")
+		}
+	} else if newMM < mongoMM {
+		if newMM != fcv {
+			return UpgradeRequest{false, "", ""}, errors.Errorf("can't upgrade to %s wits FCV set to %s", ver, fcv)
+		}
+
+		can = true
```
[the remark this answers is not in the document]
Comment (blez) [reply pkg/controller/perconaservermongodb/version.go]: https://github.com/percona/percona-server-mongodb-operator/pull/654/commits/62b940bcc932075505be5dde6851a95c951bb697
Thanks, your version is more readable
But now we have to pass only version in `Major.Minior` format into `isUpdateValid` function

**Review by dAdAbird — APPROVED** [review]

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| storage | passed |
| one-pod | passed |
| init-deploy | passed |
| upgrade-consistency | passed |
| monitoring-2-0 | passed |
| self-healing | passed |
| demand-backup | passed |
| limits | passed |
| arbiter | passed |
| scaling | passed |
| scheduled-backup | passed |
| operator-self-healing | passed |
| security-context | passed |
| service-per-pod | passed |
| demand-backup-sharded | passed |
| rs-shard-migration | passed |
| liveness | passed |
| smart-update | passed |
| version-service | passed |
| users | passed |
| upgrade | passed |
| data-sharded | passed |
| upgrade-sharded | passed |
| pitr | passed |

commit: https://github.com/percona/percona-server-mongodb-operator/pull/654/commits/62b940bcc932075505be5dde6851a95c951bb697
image: `perconalab/percona-server-mongodb-operator:PR-654-62b940bc`

**Review by qJkee — APPROVED** [review]

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| storage | passed |
| one-pod | passed |
| init-deploy | passed |
| upgrade-consistency | passed |
| self-healing | passed |
| demand-backup | passed |
| monitoring-2-0 | passed |
| limits | passed |
| scaling | passed |
| arbiter | passed |
| scheduled-backup | passed |
| security-context | passed |
| operator-self-healing | passed |
| service-per-pod | passed |
| demand-backup-sharded | passed |
| rs-shard-migration | passed |
| liveness | passed |
| smart-update | passed |
| version-service | passed |
| users | passed |
| upgrade | passed |
| data-sharded | passed |
| upgrade-sharded | passed |
| pitr | passed |

commit: https://github.com/percona/percona-server-mongodb-operator/pull/654/commits/66ea860ac4714eb30f4fc3f753cef839eca134c1
image: `perconalab/percona-server-mongodb-operator:PR-654-66ea860a`

**Review by dAdAbird — APPROVED** [review]

**Review by delgod — APPROVED** [review]

**Pull request merged**